ospfd: Show advertise router in 'show ip ospf route' command

rgwan opened the pull request:
Allows to user easily distinguish route items by router that originates it.

The output format, by example as following:

```
# show ip ospf route detail
============ OSPF network routing table ============
N T  10.0.0.0/24           [23] area: 0.0.0.0
                           via 192.168.124.67, eth7
                           adv 10.0.0.5
N    10.0.30.0/24          [24] area: 0.0.0.0
                           via 192.168.124.67, eth7
                           adv 10.0.0.5
N    10.0.50.1/32          [13] area: 0.0.0.0
                           via 192.168.124.67, eth7
                           adv 10.8.255.33
N    10.0.50.3/32          [23] area: 0.0.0.0
                           via 192.168.124.67, eth7
                           adv 10.0.0.1
N    10.0.50.7/32          [23] area: 0.0.0.0
                           via 192.168.124.67, eth7
                           adv 10.0.0.1
N T  10.0.60.0/30          [24] area: 0.0.0.0
                           via 192.168.124.67, eth7
                           adv 10.0.0.5
N    10.0.253.0/24         [34] area: 0.0.0.0
                           via 192.168.124.67, eth7
                           adv 10.0.60.2

bsp-debianrt-exp1# show ip ospf route json
{"10.0.0.0/24":{"routeType":"N","cost":32,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouterertisedRouter":"10.0.0.2"}]},"10.0.30.0/24":{"routeType":"N","cost":33,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouterertisedRouter":"10.0.0.5"}]},"10.0.50.1/32":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","z":"192.168.55.1"}]},"10.0.50.3/32":{"routeType":"N","cost":32,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.0.1"}]},"10.0.50.7/32":{"routeType":"N","cost":32,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.0.1"}]},"10.0.60.0/30":{"routeType":"N","cost":33,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.60.1"}]},"10.0.253.0/24":{"routeType":"N","cost":43,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.60.2"}]},"10.0.254.0/24":{"routeType":"N","cost":32,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.0.1"}]},"10.10.10.0/24":{"routeType":"N","cost":33,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.0.5"}]},"10.96.4.0/24":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.127.1"}]},"10.96.251.0/30":{"routeType":"N","cost":12,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.96.251.1"}]},"10.96.251.128/30":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.96.251.129"}]},"10.96.253.1/32":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"10.96.255.8/30":{"routeType":"N","cost":12,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.96.255.10"}]},"10.100.100.0/24":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.55.1"}]},"10.100.100.1/32":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.55.1"}]},"10.100.254.0/24":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.55.1"}]},"10.100.255.32/32":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.100.255.32"}]},"10.128.35.0/24":{"routeType":"N","cost":33,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.0.5"}]},"172.16.0.0/24":{"routeType":"N","cost":33,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.0.5"}]},"192.168.10.0/24":{"routeType":"N","cost":33,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"10.0.0.5"}]},"192.168.55.0/24":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.55.1"}]},"192.168.55.1/32":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.55.1"}]},"192.168.88.0/24":{"routeType":"N","cost":42,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.51.1"}]},"192.168.88.1/32":{"routeType":"N","cost":42,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.51.1"}]},"192.168.114.0/24":{"routeType":"N","cost":22,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.114.1"}]},"192.168.0.0/25":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.0.128/25":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.2.0/25":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.8.0/25":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.14.0/28":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.16.0/25":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.16.128/25":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.63.32/32":{"routeType":"N","cost":13,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.63.32"}]},"192.168.65.0/24":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"192.168.68.0/24":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"192.168.69.0/25":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"192.168.72.0/24":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"192.168.79.0/24":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"192.168.80.128/25":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"192.168.124.0/28":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.124.2"}]},"192.168.124.64/28":{"routeType":"N","cost":10,"area":"0.0.0.0","nexthops":[{"ip":" ","directlyAttachedTo":"ens192"}]},"192.168.127.1/32":{"routeType":"N","cost":10,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.127.1"}]},"192.168.127.33/32":{"routeType":"N","cost":11,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.65","via":"ens192","advertisedRouter":"192.168.127.33"}]},"192.168.127.34/32":{"routeType":"N","cost":12,"area":"0.0.0.0","nexthops":[{"ip":"192.168.124.67","via":"ens192","advertisedRouter":"192.168.127.32"}]},"192.168.127.32":{"routeType":"R ","cost":11,"area":"0.0.0.0","routerType":"asbr","nexthops":[{"ip":"192.168.124.67","via":"ens192"}]},"192.168.127.33":{"routeType":"R ","cost":10,"area":"0.0.0.0","routerType":"asbr","nexthops":[{"ip":"192.168.124.65","via":"ens192"}]},"192.168.127.36":{"routeType":"R ","cost":11,"area":"0.0.0.0","routerType":"asbr","nexthops":[{"ip":"192.168.124.67","via":"ens192"}]},"10.96.3.0/24":{"routeType":"N E2","cost":11,"type2cost":20,"tag":0,"nexthops":[{"ip":"192.168.124.65","via":"ens192"}]},"192.168.66.0/27":{"routeType":"N E2","cost":11,"type2cost":20,"tag":0,"nexthops":[{"ip":"192.168.124.65","via":"ens192"}]}}
bsp-debianrt-exp1#
```

This patch has been merged into VyOS frr fork.

ton31337 reviewed (changes requested):
Can you put all the description inside the commit? Also, we require Signed-off git tag to be used. Please follow our guidelines: https://github.com/FRRouting/frr/blob/master/doc/developer/workflow.rst#submitting-patches-and-enhancements

P.S. Looks like tests must be adopted as well for this new output.

rgwan commented:
> Can you put all the description inside the commit? Also, we require Signed-off git tag to be used. Please follow our guidelines: https://github.com/FRRouting/frr/blob/master/doc/developer/workflow.rst#submitting-patches-and-enhancements
> 
> P.S. Looks like tests must be adopted as well for this new output.

In order to not break current tests, I changed this patch, add a argument named 'verbose' to enable the verbose output for showing advertiser router.And I added a sign-off in this commit.

ton31337 commented on ospfd/ospf_vty.c:
```diff
@@ -11145,13 +11165,14 @@ static int show_ip_ospf_route_common(struct vty *vty, struct ospf *ospf,
 	VRF_CMD_HELP_STR
 	"All VRFs\n"
 	"OSPF routing table\n"
+	"Verbose output\n"
```
`Detail Information`

rgwan replied:
Fixed

ton31337 commented on ospfd/ospf_vty.c:
```diff
@@ -11163,11 +11184,16 @@ DEFUN (show_ip_ospf_route,
 	if (uj)
 		json = json_object_new_object();
 
+	if (argv[argc - (uj == true? 2: 1)]->arg &&
```
Convert this function DEFUN to DEFPY and you will be able to avoid this _funky_ logic here and below. See for more examples like (`debug_ospf_packet_cmd` or so).

rgwan replied:
I choose to use `argv_find()` to parse user's input like `show_ip_ospf_gr_helper()` do.

ton31337 commented on ospfd/ospf_vty.c:
```diff
@@ -11235,16 +11261,21 @@ DEFUN (show_ip_ospf_route,
 	unsigned short instance = 0;
+	bool verbose = false;
+
+	if (argv[argc - 1]->arg && strmatch(argv[argc - 1]->text, "verbose"))
```
Ditto.

rgwan replied:
Fixed

ton31337 reviewed (changes requested):
If we add a new CLI command/option, we require a documentation update, please update it. Also, I believe this MUST be adopted for OSPFv3.

rgwan commented on ospfd/ospf_vty.c:
```diff
@@ -10401,7 +10401,7 @@ static void config_write_stub_router(struct vty *vty, struct ospf *ospf)
 static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 				       struct route_table *rt,
-				       json_object *json)
+				       json_object *json, uint8_t verbose)
```
Changed `verbose` into `detail`, fixed data type.

rgwan commented:
> If we add a new CLI command/option, we require a documentation update, please update it. Also, I believe this MUST be adopted for OSPFv3.

I updated OSPF documentation, but it seems OSPFv3 has already supported it.

`show ipv6 ospf6 [vrf <NAME|all>] route [<intra-area|inter-area|external-1|external-2|X:X::X:X|X:X::X:X/M|detail|summary>] [json]`

ton31337 commented:
@frrbot rereview

ton31337 commented:
@rgwan please apply frrbot styling issues except for DEFUN. https://github.com/FRRouting/frr/pull/13140/checks?check_run_id=12437161091

diodep commented:
> @rgwan please apply frrbot styling issues except for DEFUN. https://github.com/FRRouting/frr/pull/13140/checks?check_run_id=12437161091

Applied.

ton31337 reviewed (approved):
LGTM

rgwan commented:
> LGTM

Okay, could you please merge it?

rgwan commented:
The newly added feature can distinguish transit or stub network in OSPF intra area. Example output as following:

```
bsp-debianrt-exp1# show ip ospf route detail 
============ OSPF network routing table ============
N    10.0.0.0/24           [32] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5, transit
N    10.0.30.0/24          [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5, stub
N    10.0.50.1/32          [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.33, stub
N    10.0.50.3/32          [32] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.1, stub
N    10.0.50.7/32          [32] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.1, stub
N    10.0.60.0/30          [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5, transit
N    10.0.253.0/24         [43] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.60.2, stub
N    10.0.254.0/24         [32] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.1, stub
N    10.2.1.0/24           [24] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.64, stub
N    10.2.2.0/24           [24] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.64, stub
N    10.2.24.0/24          [24] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.64, stub
N    10.2.192.0/28         [23] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.32, transit
N    10.2.192.32/28        [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.32, transit
N    10.2.224.0/24         [24] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.64, stub
N    10.2.254.0/24         [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.33, stub
N    10.2.255.32/32        [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.32, stub
N    10.2.255.33/32        [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.33, stub
N    10.2.255.64/32        [23] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.64, stub
N    10.10.10.0/24         [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5, stub
N    10.96.4.0/24          [11] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.127.1, stub
N    10.96.251.0/30        [12] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.127.36, transit
N    10.96.251.128/30      [11] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.127.36, transit
N    10.96.255.8/30        [12] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, transit
N    10.100.100.0/24       [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.33, stub
N    10.100.100.1/32       [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.33, stub
N    10.128.35.0/24        [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5, stub
N    172.16.0.0/24         [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5, stub
N    192.168.10.0/24       [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5, stub
N    192.168.55.0/24       [22] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.2.255.33, stub
N    192.168.88.0/24       [42] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.51.1, stub
N    192.168.88.1/32       [42] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.51.1, stub
N    192.168.0.0/25        [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.0.128/25      [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.2.0/25        [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.8.0/25        [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.14.0/28       [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.16.0/25       [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.16.128/25     [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.63.32/32      [13] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.63.32, stub
N    192.168.65.0/24       [11] area: 0.0.0.0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33, stub
N    192.168.68.0/24       [11] area: 0.0.0.0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33, stub
N    192.168.69.0/25       [11] area: 0.0.0.0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33, stub
N    192.168.72.0/24       [11] area: 0.0.0.0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33, stub
N    192.168.79.0/24       [11] area: 0.0.0.0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33, stub
N    192.168.80.128/25     [11] area: 0.0.0.0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33, stub
N    192.168.124.0/28      [11] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.127.1, transit
N    192.168.124.64/28     [10] area: 0.0.0.0
                           directly attached to ens192
N    192.168.127.1/32      [10] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.127.1, stub
N    192.168.127.33/32     [11] area: 0.0.0.0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33, stub
N    192.168.127.34/32     [12] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 192.168.127.32, stub
N    192.168.127.64/32     [10] area: 0.0.0.0
                           directly attached to dummy0

============ OSPF router routing table =============
R    192.168.127.32         [11] area: 0.0.0.0, ASBR
                           via 192.168.124.67, ens192
R    192.168.127.33         [10] area: 0.0.0.0, ASBR
                           via 192.168.124.65, ens192
R    192.168.127.36         [11] area: 0.0.0.0, ASBR
                           via 192.168.124.67, ens192

============ OSPF external routing table ===========
N E2 10.96.3.0/24          [11/20] tag: 0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33
N E2 192.168.66.0/27       [11/20] tag: 0
                           via 192.168.124.65, ens192
                           adv 192.168.127.33


bsp-debianrt-exp1# 
```


Reference output as following:

```
<hpe-r1>disp ospf routing  

OSPF Process 1 with Router ID 10.2.255.64

 Routing for Network
 ------------------------------------------------------------------------------
 Destination            Cost Type       Next-Hop        AdvRouter       Area           
 10.0.0.0/24              22 Transit    10.2.192.1      10.0.0.5        0.0.0.0        
 10.0.30.0/24             23 Stub       10.2.192.1      10.0.0.5        0.0.0.0        
 10.0.50.1/32             12 Stub       10.2.192.1      10.2.255.33     0.0.0.0        
 10.0.50.3/32             22 Stub       10.2.192.1      10.0.0.1        0.0.0.0        
 10.0.50.7/32             22 Stub       10.2.192.1      10.0.0.1        0.0.0.0        
 10.0.60.0/30             23 Transit    10.2.192.1      10.0.0.5        0.0.0.0        
 10.0.253.0/24            33 Stub       10.2.192.1      10.0.60.2       0.0.0.0        
 10.0.254.0/24            22 Stub       10.2.192.1      10.0.0.1        0.0.0.0        
 10.2.1.0/24               1 Direct     10.2.1.1        10.2.255.64     0.0.0.0        
 10.2.2.0/24               1 Direct     10.2.2.1        10.2.255.64     0.0.0.0        
 10.2.24.0/24              1 Direct     10.2.24.1       10.2.255.64     0.0.0.0        
 10.2.192.0/28             1 Direct     10.2.192.2      10.2.255.64     0.0.0.0        
 10.2.192.32/28            2 Transit    10.2.192.1      10.2.255.32     0.0.0.0        
 10.2.224.0/24             1 Direct     10.2.224.1      10.2.255.64     0.0.0.0        
 10.2.254.0/24            12 Stub       10.2.192.1      10.2.255.33     0.0.0.0        
 10.2.255.32/32            1 Stub       10.2.192.1      10.2.255.32     0.0.0.0        
 10.2.255.33/32           12 Stub       10.2.192.1      10.2.255.33     0.0.0.0        
 10.2.255.64/32            0 Direct     10.2.255.64     10.2.255.64     0.0.0.0        
 10.10.10.0/24            23 Stub       10.2.192.1      10.0.0.5        0.0.0.0        
 10.96.4.0/24             14 Stub       10.2.192.1      192.168.127.1   0.0.0.0        
 10.96.251.0/30           12 Transit    10.2.192.1      192.168.127.36  0.0.0.0        
 10.96.251.128/30         13 Transit    10.2.192.1      192.168.127.36  0.0.0.0        
 10.96.255.8/30           15 Transit    10.2.192.1      192.168.63.32   0.0.0.0        
 10.100.100.0/24          12 Stub       10.2.192.1      10.2.255.33     0.0.0.0        
 10.100.100.1/32          12 Stub       10.2.192.1      10.2.255.33     0.0.0.0        
 10.128.35.0/24           23 Stub       10.2.192.1      10.0.0.5        0.0.0.0        
 172.16.0.0/24            23 Stub       10.2.192.1      10.0.0.5        0.0.0.0        
 192.168.10.0/24          23 Stub       10.2.192.1      10.0.0.5        0.0.0.0        
 192.168.55.0/24          12 Stub       10.2.192.1      10.2.255.33     0.0.0.0        
 192.168.88.0/24          32 Stub       10.2.192.1      192.168.51.1    0.0.0.0        
 192.168.88.1/32          32 Stub       10.2.192.1      192.168.51.1    0.0.0.0        
 192.168.0.0/25           16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.0.128/25         16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.2.0/25           16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.8.0/25           16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.14.0/28          16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.16.0/25          16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.16.128/25        16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.63.32/32         16 Stub       10.2.192.1      192.168.63.32   0.0.0.0        
 192.168.65.0/24          15 Stub       10.2.192.1      192.168.127.33  0.0.0.0        
 192.168.68.0/24          15 Stub       10.2.192.1      192.168.127.33  0.0.0.0        
 192.168.69.0/25          15 Stub       10.2.192.1      192.168.127.33  0.0.0.0        
 192.168.72.0/24          15 Stub       10.2.192.1      192.168.127.33  0.0.0.0        
 192.168.79.0/24          15 Stub       10.2.192.1      192.168.127.33  0.0.0.0        
 192.168.80.128/25        15 Stub       10.2.192.1      192.168.127.33  0.0.0.0        
 192.168.124.0/28         14 Transit    10.2.192.1      192.168.127.1   0.0.0.0        
 192.168.124.64/28        14 Transit    10.2.192.1      192.168.127.1   0.0.0.0        
 192.168.127.1/32         13 Stub       10.2.192.1      192.168.127.1   0.0.0.0        
 192.168.127.33/32        15 Stub       10.2.192.1      192.168.127.33  0.0.0.0        
 192.168.127.34/32        15 Stub       10.2.192.1      192.168.127.32  0.0.0.0        
 192.168.127.64/32        24 Stub       10.2.192.1      192.168.127.64  0.0.0.0        

 Routing for ASEs
 ------------------------------------------------------------------------------
 Destination              Cost Type              Tag Next-Hop        AdvRouter      
 10.96.3.0/24               20 Type2               0 10.2.192.1      192.168.127.33  
 192.168.66.0/27            20 Type2               0 10.2.192.1      192.168.127.33  

 Total Nets: 53
 Intra Area: 51  Inter Area: 0  ASE: 2  NSSA: 0
<hpe-r1>

```

ton31337 commented on ospfd/ospf_vty.c:
```diff
@@ -10526,6 +10526,10 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 									ospf->vrf_id));
+							json_object_string_addf(
+								json_nexthop,
+								"adv", "%pI4",
```
I suggest a more descriptive JSON key name...

rgwan replied:
I have changed it to 'advertiser'.

rgwan commented:
@ton31337 What do you think? Is it time to merge this?

ton31337 reviewed (approved):
LGTM

rgwan commented:
> LGTM

Does this system automatically merge this after all test passed? Some topo test seems not very stable, failing randomly.....

Is that means in some case maintainers need to merge it manually?

ton31337 commented:
Someone from the maintainers has to click the merge button. But I'll also keep this open to someone else from maintainers to review.

diodep commented:
> Someone from the maintainers has to click the merge button. But I'll also keep this open to someone else from maintainers to review.

Btw, which display style do you prefer in these two?
```
N T  10.0.0.0/24           [32] area: 0.0.0.0
 adv 10.0.0.5              via 198.18.124.67, ens192
N    10.0.30.0/24          [33] area: 0.0.0.0
 adv 10.0.0.5              via 198.18.124.67, ens192
N    10.0.50.1/32          [22] area: 0.0.0.0

```
```
N T  10.0.0.0/24           [32] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5
N    10.0.30.0/24          [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5
```

mwinter-osr commented:
CI:rerun ARM testbeds on part 4 & 5 have currently a known issue. Rerunning with these disabled

rgwan commented:
> > Someone from the maintainers has to click the merge button. But I'll also keep this open to someone else from maintainers to review.
> 
> Btw, which display style do you prefer in these two?
> 
> ```
> N T  10.0.0.0/24           [32] area: 0.0.0.0
>  adv 10.0.0.5              via 198.18.124.67, ens192
> N    10.0.30.0/24          [33] area: 0.0.0.0
>  adv 10.0.0.5              via 198.18.124.67, ens192
> N    10.0.50.1/32          [22] area: 0.0.0.0
> ```
> 
> ```
> N T  10.0.0.0/24           [32] area: 0.0.0.0
>                            via 192.168.124.67, ens192
>                            adv 10.0.0.5
> N    10.0.30.0/24          [33] area: 0.0.0.0
>                            via 192.168.124.67, ens192
>                            adv 10.0.0.5
> ```

I prefer the first one but the second one is more compact on small display.

rgwan commented:
@Jafaral Do you think it is good to be merged in mainline?

rgwan commented:
@ton31337 I updated example output format in the top post. Are you still working on this?

rgwan commented:
@ton31337 It's already June....

mobash-rasool commented on doc/user/ospfd.rst:
```diff
@@ -824,10 +824,11 @@ Showing Information
 
    Show the OSPF routing table, as determined by the most recent SPF
-   calculation.
+   calculation. If detail is specified, each routing item's
```
Can you please correct the last sentence here. It is not clear what will be displayed if detail is specified in this statement. It will be better if you can mention it a little more clearly.

mobash-rasool replied:
You can mention something like, detail option can be used for more information like advertising router ID for the route, etc.

rgwan replied:
OK.

mobash-rasool commented on ospfd/ospf_route.h:
```diff
@@ -69,6 +69,8 @@ struct route_standard {
 	/*  */
 	uint8_t flags; /* From router-LSA */
+
+	bool transit;
```
Do we need this field to be added? Isn't the fields path_type as OSPF_PATH_INTRA_AREA and type as OSPF_DESTINATION_NETWORK means it is a transit route?

rgwan replied:
I'm sorry that I'm not so familiar with OSPF protocol. Fixed.

mobash-rasool commented on ospfd/ospf_vty.c:
```diff
@@ -10793,6 +10804,11 @@ static void show_ip_ospf_route_external(struct vty *vty, struct ospf *ospf,
 								ospf->vrf_id));
+						json_object_string_addf(
+							json_nexthop,
+							"advertisedRouter",
```
You are referring to something which is already wrong.
Please refer to the right ones like Line 6326 and 7058.
When you say advertisingRouter- It means this is the router which has advertised the LSA.
But if you say advertisedRouter - It means this router has been advertised which is not what you mean in the output above.

rgwan replied:
> Please refer to the right ones like Line 6326 and 7058.

Well, my local version it's at line 6090, branch 'master' it's at line 6271:

```
						&lsa->data->id);
			json_object_string_addf(json_lsa, "advertisedRouter",
						"%pI4", &lsa->data->adv_router);
			json_object_int_add(json_lsa, "lsaAge", LS_AGE(lsa));
			json_object_string_add(json_lsa, "sequenceNumber",
```

mobash-rasool commented:
@rgwan: Please rebase you changes.

mobash-rasool commented:
Your changes are causing the topotest failure. Please check.

diodep commented:
> Your changes are causing the topotest failure. Please check.
```

AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
   Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
       Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
    assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
```

I haven't touch OSPFv3 at all, this fail is unlikely caused by my changes.

rgwan commented:
> Your changes are causing the topotest failure. Please check.
```

AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
   Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
       Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
    assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
```

I haven't touch OSPFv3 at all, this fail is unlikely caused by my changes.

mobash-rasool commented:
> > Your changes are causing the topotest failure. Please check.
> 
> ```
> 
> AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
>    Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
> assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
>        Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
>     assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> ```
> 
> I haven't touch OSPFv3 at all, this fail is unlikely caused by my changes.

Check this link for more details:
https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239

Also look at below outputs:
test_ospf_gr_topo1 [test_initial_convergence](https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239/test/case/109024286)
AssertionError: "rt1" JSON output mismatches the expected result
assert Generated JSON diff error report:
  
  > $->1.1.1.1/32->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
  > $->10.0.1.0/24->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
E   AssertionError: "rt1" JSON output mismatches the expected result
    assert Generated JSON diff error report:
      
(2 more lines...)

diodep commented:
> > > Your changes are causing the topotest failure. Please check.
> > 
> > 
> > ```
> > 
> > AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
> >    Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
> > assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> > E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
> >        Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
> >     assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I haven't touch OSPFv3 at all, this fail is unlikely caused by my changes.
> 
> Check this link for more details: https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239
> 
> Also look at below outputs: test_ospf_gr_topo1 [test_initial_convergence](https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239/test/case/109024286) AssertionError: "rt1" JSON output mismatches the expected result assert Generated JSON diff error report:
> 
> > $->1.1.1.1/32->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
> > $->10.0.1.0/24->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
> > E   AssertionError: "rt1" JSON output mismatches the expected result
> > assert Generated JSON diff error report:
> 
> (2 more lines...)

Updated, use old JSON format when 'detail' option is not used.

rgwan commented:
> > > Your changes are causing the topotest failure. Please check.
> > 
> > 
> > ```
> > 
> > AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
> >    Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
> > assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> > E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
> >        Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
> >     assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I haven't touch OSPFv3 at all, this fail is unlikely caused by my changes.
> 
> Check this link for more details: https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239
> 
> Also look at below outputs: test_ospf_gr_topo1 [test_initial_convergence](https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239/test/case/109024286) AssertionError: "rt1" JSON output mismatches the expected result assert Generated JSON diff error report:
> 
> > $->1.1.1.1/32->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
> > $->10.0.1.0/24->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
> > E   AssertionError: "rt1" JSON output mismatches the expected result
> > assert Generated JSON diff error report:
> 
> (2 more lines...)

Updated, use old JSON format when 'detail' option is not used.

rgwan commented:
> > > Your changes are causing the topotest failure. Please check.
> > 
> > 
> > ```
> > 
> > AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
> >    Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
> > assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> > E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 : Failed 
> >        Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
> >     assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I haven't touch OSPFv3 at all, this fail is unlikely caused by my changes.
> 
> Check this link for more details: https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239
> 
> Also look at below outputs: test_ospf_gr_topo1 [test_initial_convergence](https://ci1.netdef.org/browse/FRR-PULLREQ2-TOPO9U18AMD64-12239/test/case/109024286) AssertionError: "rt1" JSON output mismatches the expected result assert Generated JSON diff error report:
> 
> > $->1.1.1.1/32->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
> > $->10.0.1.0/24->routeType: d1 has element with value 'N T' but in d2 it has value 'N'
> > E   AssertionError: "rt1" JSON output mismatches the expected result
> > assert Generated JSON diff error report:
> 
> (2 more lines...)

Now everything is clear.

rgwan commented:
@mobash-rasool Do you have anymore comment on this pull request? Is it ok to merge?

diodep commented:
@ton31337 Do we have a chance to merge this patch in this month?

ton31337 commented on ospfd/ospf_vty.c:
```diff
@@ -10724,18 +10725,21 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
-					buf1, or->cost,
-					&or->u.std.area_id);
+				vty_out(vty, "N %s  %-18s    [%d] area: %pI4\n",
+					or_is_transit && detail ? "T" : " ",
```
Maybe we can give a more descriptive information (T - transitive, N - ...) into a header or so?

rgwan replied:
> Maybe we can give a more descriptive information (T - transitive, N - ...) into a header or so?

like this?

```
bsp-debianrt-exp1# show ip ospf route detail
============ OSPF network routing table ============
N T  10.0.0.0/24           [32] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5
N    10.0.30.0/24          [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5
...
============ OSPF external routing table ===========
N - network         T - transitive  IA - intra-area
E - external route  D - destination R  - router
```

ton31337 replied:
Yes, but at the top of the output :)

diodep replied:
> Yes, but at the top of the output :)

Ok, I change the output format it into this.

```
bsp-debianrt-exp1# show ip ospf rou de
N - network         T - transitive  IA - intra-area
E - external route  D - destination R  - router

============ OSPF network routing table ============
N T  10.0.0.0/24           [32] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5
N T  10.0.30.0/24          [33] area: 0.0.0.0
                           via 192.168.124.67, ens192
                           adv 10.0.0.5
```

ton31337 reviewed (approved):
LGTM, just not sure 100% about this commit https://github.com/FRRouting/frr/pull/13140/commits/3940408fe90cc96d069233761159330253ee5dae. Maybe someone with a better understanding of OSPF can review it here: @rwestphal @odd22? 

Other than that, can you fix frrbot styling issues (except DEFUN) and add an updated version of CLI outputs into commits inside (not PR)?

mobash-rasool commented:
> LGTM, just not sure 100% about this commit [3940408](https://github.com/FRRouting/frr/commit/3940408fe90cc96d069233761159330253ee5dae). Maybe someone with a better understanding of OSPF can review it here: @rwestphal @odd22?
> 
> Other than that, can you fix frrbot styling issues (except DEFUN) and add an updated version of CLI outputs into commits inside (not PR)?
> 
> Also, maybe it would be worth adding a header for the detail command (that displays flags T/N/...).

@ton31337 : This commit https://github.com/FRRouting/frr/commit/3940408fe90cc96d069233761159330253ee5dae is right. Vertex ID is not same as advertising router in some LSAs.

I have a concern, this PR is changing the existing CLI output as well which may impact people using screen scraping the CLIs in their test scripts. Is that okay?

ton31337 commented:
>I have a concern, this PR is changing the existing CLI output as well which may impact people using screen scraping the CLIs in their test scripts. Is that okay?

IMO this is okay for non-JSON outputs :shrug: Let's ask others too. /cc @donaldsharp

rgwan commented:
> > I have a concern, this PR is changing the existing CLI output as well which may impact people using screen scraping the CLIs in their test scripts. Is that okay?
> 
> IMO this is okay for non-JSON outputs 🤷 Let's ask others too. /cc @donaldsharp

When the 'detail' flag is not used in command line, it will use old CLI output format to not break any test scripts relys on CLI output.

mobash-rasool commented:
> > > I have a concern, this PR is changing the existing CLI output as well which may impact people using screen scraping the CLIs in their test scripts. Is that okay?
> > 
> > 
> > IMO this is okay for non-JSON outputs 🤷 Let's ask others too. /cc @donaldsharp
> 
> When the 'detail' flag is not used in command line, it will use old CLI output format to not break any test scripts relys on CLI output.

You have modified show_ip_ospf_route_network api which is being used in few of the existing CLIs, please check the callers.

rgwan commented:
> > > > I have a concern, this PR is changing the existing CLI output as well which may impact people using screen scraping the CLIs in their test scripts. Is that okay?
> > > 
> > > 
> > > IMO this is okay for non-JSON outputs 🤷 Let's ask others too. /cc @donaldsharp
> > 
> > 
> > When the 'detail' flag is not used in command line, it will use old CLI output format to not break any test scripts relys on CLI output.
> 
> You have modified show_ip_ospf_route_network api which is being used in few of the existing CLIs, please check the callers.

It seems `show_ip_ospf_route_network` only used by function `show_ip_ospf_route_common`, I think it's safe to do this.

```
dev@bsp-dev1:~/dev/frr$ grep -R "show_ip_ospf_route_network"
ospfd/ospf_vty.c:static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
ospfd/ospf_vty.c:       show_ip_ospf_route_network (vty, ospf->new_table);   */ # just a comment
ospfd/ospf_vty.c:       show_ip_ospf_route_network(vty, ospf, ospf->new_table, json_vrf,
```

rgwan commented:
In order to keep compatibility for JSON output, I add a new attribute named 'transit', and it will be set true in a transit network item.

@mobash-rasool Besides, path_type == OSPF_PATH_INTRA_AREA and type == OSPF_DESTINATION_NETWORK also can means it is a stub route. I reverted change and add back `transit` attribute for struct `route_standard`.

rgwan commented:
@mobash-rasool @ton31337 I think it's good to merge?

ton31337 commented:
Probably OK, but I'd slightly modify the header to be (to be more or less similar to `show ip route`, `show ip bgp`, etc.):

```
bsp-debianrt-exp1# show ip ospf route detail
Codes: N - network, T - transitive, IA - inter-area, 
       E - external route, D - destination, R  - router
```

rgwan commented:
> Probably OK, but I'd slightly modify the header to be (to be more or less similar to `show ip route`, `show ip bgp`, etc.):
> 
> ```
> bsp-debianrt-exp1# show ip ospf route detail
> Codes: N - network, T - transitive, IA - inter-area, 
>        E - external route, D - destination, R  - router
> ```

It's done.

rgwan commented:
> Probably OK, but I'd slightly modify the header to be (to be more or less similar to `show ip route`, `show ip bgp`, etc.):
> 
> ```
> bsp-debianrt-exp1# show ip ospf route detail
> Codes: N - network, T - transitive, IA - inter-area, 
>        E - external route, D - destination, R  - router
> ```

Is that good time to merge it?

ton31337 commented on ospfd/ospf_vty.c:
```diff
@@ -10727,15 +10727,19 @@ static void show_ip_ospf_route_network(struct vty *vty, struct ospf *ospf,
 			if (json) {
 				json_object_string_add(json_route, "routeType",
 						       "N");
+				json_object_string_add(json_route, "transit",
```
`json_object_bool_add(json_route, "transit", or->u.std.transit);`

ton31337 reviewed (approved):
One nit and LGTM.

ton31337 commented:
Also, can you take a look at these CI warnings?

```
<stdin>:14: trailing whitespace.
   calculation. When detail option is used, it shows more information 
warning: 1 line adds whitespace errors.
Report for ospf_vty.c | 2 issues
===============================================
< WARNING: Too many leading tabs - consider code refactoring
< #10813: FILE: /tmp/f1-1862051/ospf_vty.c:10813:
```

odd22 commented:
> LGTM, just not sure 100% about this commit [3940408](https://github.com/FRRouting/frr/commit/3940408fe90cc96d069233761159330253ee5dae). Maybe someone with a better understanding of OSPF can review it here: @rwestphal @odd22?
> 
> Other than that, can you fix frrbot styling issues (except DEFUN) and add an updated version of CLI outputs into commits inside (not PR)?
> 
> Also, maybe it would be worth adding a header for the detail command (that displays flags T/N/...).

Commit is right. the Vertex ID is fulfill with the LSA ID which is different from the Router Advertisement, even if it is also an in_addr_t.

rgwan commented:
> Also, can you take a look at these CI warnings?
> 
> ```
> <stdin>:14: trailing whitespace.
>    calculation. When detail option is used, it shows more information 
> warning: 1 line adds whitespace errors.
> Report for ospf_vty.c | 2 issues
> ===============================================
> < WARNING: Too many leading tabs - consider code refactoring
> < #10813: FILE: /tmp/f1-1862051/ospf_vty.c:10813:
> ```

Doc fixed and use `json_object_bool_add` to add `transit` attribute.

rgwan commented:
> One nit and LGTM.

Is it good time to merge it?

rgwan commented:
@ton31337 @odd22 Another week just passed....

ton31337 commented:
@rgwan sorry about the delay... Waiting for CI, and gonna merge this.

rgwan commented:
> # Continuous Integration Result: SUCCESSFUL
> 
> Congratulations, this patch passed basic tests
> 
> Tested-by: NetDEF / OpenSourceRouting.org CI System
> 
> CI System Testrun URL: https://ci1.netdef.org/browse/FRR-PULLREQ2-12863/
> 
> This is a comment from an automated CI system. For questions and feedback in regards to this CI system, please feel free to email Martin Winter - mwinter (at) opensourcerouting.org.
> ### Warnings Generated during build:
> Checkout code: Successful with additional warnings

Patch itself seems good, and I rebased it to latest commit and it fails @ton31337

ton31337 commented:
All good, I submitted to rerun only failed tests, let's see in a couple of minutes.
